### ISSUE TITLE
static binary causes a segmentation fault

### ISSUE THREAD

**Issue by soonhokong** [leanprover/lean3#557]
@leodemoura, could you take a look at this when you have some time?
## Context

There is a student in the class who is using Debian7 (wheezy). He asked @avigad and me how to install Lean on his system. First, I looked up how to install gcc-4.8 or higher on Debian7. I found an [article](http://unix.stackexchange.com/questions/89243/how-to-install-gcc-4-8-1-on-debian-wheezy) explaining how to do it, but it seems that it's not a good idea to ask other people to follow this.

Then I thought about making a static binary and wrapping the generated binary and compiled libraries as a debian package (`.deb`). I know that Ubuntu-12.04 and Debian wheezy are compatible so the student should be able to install this `.deb` file with `dpkg -i` command.

I made the following changes in my [ppa-updater](https://github.com/soonhokong/lean-ppa-updater) repository to drop out dependencies on libstdc++/libgmp/libmpfr/liblua and to add `-DSTATIC=ON`.
- https://github.com/soonhokong/lean-ppa-updater/commit/af60731f550745c83265800ed0a634e7f89531fd
- https://github.com/soonhokong/lean-ppa-updater/commit/862ba3460c476709a2a12caa0a2d3f30cc996352
## Problem

After pushing the changes to the launchpad, I got a build error. Here is a trace: https://launchpadlibrarian.net/204096875/buildlog_ubuntu-precise-amd64.lean_0.2.0.20150423102844.git2613e7c4443e4c38c177260a2e4be5fa2e4c86ec~12.04_BUILDING.txt.gz
1. When the HOTT library is compiled, lean generates a segmentation fault.
2. I'm not sure that it's related, but It generates the following warning when building a static binary:

```
(.text+0x816): warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
make[3]: Leaving directory `/build/buildd/lean-0.2.0.20150423102844.git2613e7c4443e4c38c177260a2e4be5fa2e4c86ec~12.04/build_dir'
```

I was able to reproduce the problem on my local machine as well. Here is a [valgrind trace](https://gist.github.com/soonhokong/2be71ef14100509b0ccd).

**Comment by leodemoura** [comment]
I think the problem is pthreads.
See this link:
http://stackoverflow.com/questions/10390584/force-gcc-to-static-link-e-g-pthreads-and-not-dynamic-link

We can confirm that by compiling Lean without multithreading support.

**Comment by soonhokong** [comment]
I can confirm that using `-DMULTI_THREAD=OFF` fixes the problem.

**Comment by soonhokong** [comment]
I found another link https://gcc.gnu.org/ml/gcc-help/2010-05/msg00029.html:

> Another fix is to use -Wl,--whole-archive -lpthread -Wl,--no-whole-archive when linking.

Manually adding them to https://github.com/leanprover/lean/blob/master/src/CMakeLists.txt#L91 solves the problem. But it generates the following warning:

```
sem_open.c:333: warning: the use of `mktemp' is dangerous, better use `mkstemp' or `mkdtemp'
```

**Comment by leodemoura** [comment]
I'm confused by the warning. When is it generated? 
BTW, `sem_open.c` is not part of Lean.

**Comment by soonhokong** [comment]
> When is it generated? 

When I build Lean (compile-time).

> BTW, sem_open.c is not part of Lean.

You're right. It's part of glibc code https://sourceware.org/bugzilla/show_bug.cgi?id=15137

**Comment by leodemoura** [comment]
@soonhokong I added the flag you suggested. It seems to work on my machine.
Could you check on yours?

Thanks

**Comment by soonhokong** [comment]
@leodemoura: sure, I will check it tonight. I'll also commit the changes that I made in `CMakeLists.txt` to generate a Debian package using `cpack -G Deb`. I used that to generate https://github.com/leanprover/bin/blob/master/lean-0.2.0-debian.deb.

**Comment by soonhokong** [comment]
@leodemoura, it worked on my machine. I also added Debian support for cpack in e8d894473ed8d57cdf2616e6aa1b7df1c589a6ca. I checked that a generated `.deb` file can be used in Debian-6/7/8 without installing any dependencies (by default, they have python installed).

It would be great if you can set-up a virtual machine which builds Lean, run "cpack -G DEB", and upload the generated to https://github.com/leanprover/bin. I've updated the instruction at the webpage, http://leanprover.github.io/download/debian7.html, to find `.deb` file at https://github.com/leanprover/bin.

**Comment by leodemoura** [comment]
Great! Many thanks. 
I will install the VM.

**Comment by soonhokong** [comment]
Thanks!!
- I've updated the download instructions for Debian to include Debian-8. Now we support Debian-7 and Debian-8.
- I've changed ppa-updater to use `-DSTATIC=ON`, triggered PPA updates on launchpad. [Everything is good now](https://launchpad.net/~leanprover/+archive/ubuntu/lean/).

**Comment by leodemoura** [comment]
So, is the binary for Ubuntu (in the PPA) also `static` now?

**Comment by soonhokong** [comment]
Yes, mainly I wanted to check whether it's working or not. Do you see a problem there? We can revert to `-DSTATIC=OFF`.

**Comment by leodemoura** [comment]
> Do you see a problem there?

I have only random observations:
- The binary is bigger.
- Most binaries installed using PPA are not static, right? Systems like PPA were designed to make sure we can install dynamic binaries in a convenient way. 
- We test and use the dynamic version. So, if there is a problem with the static version, it may take a long time to find out (probably a user will find the problem, and notify us).
- For Debian, it is fine to use the static, since it is not one of the main platforms we have nice support (OSX, Windows, and Ubuntu).
- I got a bunch of warnings when I compiled using `-DSTATIC=ON` on my machine.
